src/backends/perf_pt/perf_pt.c

```diff
@@ -348,124 +318,173 @@ tracer_thread(void *arg)
         failing = true;
         goto clean;
     }
-    tr_ctx->stop_fd_rd = stop_fds[0];
-    tr_ctx->stop_fd_wr = stop_fds[1];
 
-    // Allocate buffer to copy the trace into (currently the same size as the AUX buffer).
-    __u64 trace_bufsize = tr_conf->aux_bufsize * getpagesize();
+clean:
+    if (failing) {
+        perf_pt_free_tracer(tr_ctx);
+        return NULL;
+    }
+    return tr_ctx;
+}
+
+/*
+ * Turn on Intel PT.
+ *
+ * Returns zero on success, or -1 otherwise.
+ */
+int
+perf_pt_start_tracer(struct tracer_ctx *tr_ctx)
+{
+    int clean_sem = 0, clean_thread = 0;
+    int ret = 0;
+
+    // A pipe to signal the trace thread to stop.
+    //
+    // It has to be a pipe becuase it needs to be used in a poll(6) loop later.
+    if (pipe(tr_ctx->stop_fds) != 0) {
+        ret = -1;
+        goto clean;
+    }
+
+    // Allocate buffer to copy the trace into (currently the same size as the
+    // AUX buffer). We cannot re-use the same buffer for different calls to
+    // perf_pt_start_tracer(), since the buffers are stored on the Rust side.
+    __u64 trace_bufsize = tr_ctx->aux_bufsize * getpagesize();
     tr_ctx->trace_buf = malloc(trace_bufsize);
     if (tr_ctx->trace_buf == NULL) {
-        failing = true;
+        ret = -1;
         goto clean;
     }
 
     // Use a semaphore to wait for the tracer to be ready.
     sem_t tracer_init_sem;
     if (sem_init(&tracer_init_sem, 0, 0) == -1) {
-        failing = true;
+        ret = -1;
         goto clean;
     }
     clean_sem = 1;
 
     // Build the arguments struct for the tracer thread.
     struct tracer_thread_args thr_args = {
-        perf_fd,
-        stop_fds[0],
+        tr_ctx->perf_fd,
+        tr_ctx->stop_fds[0],
         &tracer_init_sem,
-        tr_conf->data_bufsize,
-        tr_conf->aux_bufsize,
         &tr_ctx->trace_buf,
         trace_bufsize,
         &tr_ctx->trace_len,
+        tr_ctx->aux_buf,
+        tr_ctx->base_buf, // The header is the first region in the base buf.
     };
 
     // Spawn a thread to deal with copying out of the PT AUX buffer.
     int rc = pthread_create(&tr_ctx->tracer_thread, NULL, tracer_thread, &thr_args);
     if (rc) {
-        failing = true;
+        ret = -1;
         goto clean;
     }
+    clean_thread = 1;
 
     // Wait for the tracer to initialise, and check it didn't fail.
     rc = -1;
     while (rc == -1) {
         rc = sem_wait(&tracer_init_sem);
         if ((rc == -1) && (errno != EINTR)) {
-            failing = true;
+            ret = -1;
             goto clean;
         }
     }
 
     // Turn on tracing hardware.
     if (ioctl(tr_ctx->perf_fd, PERF_EVENT_IOC_ENABLE, 0) < 0) {
-        failing = true;
+        ret = -1;
         goto clean;
     }
 
-    ret = tr_ctx;
 clean:
-    if (failing) {
-        ret = NULL;
-        // Child thread must be stopped before closing the shared perf_fd.
+    if (clean_sem) {
+        if (sem_destroy(&tracer_init_sem) == -1) {
```

Comment:
Can be folded with `&&` into the outer `if`.

Reply:
True, will fix.